### PR TITLE
nuclei parser and unittests improvements

### PR DESCRIPTION
**Pull request by kir-b**
**Description**

This PR improves Nuclei parser.

- Adds import more details from Nuclei finding, i.e. tags, request, response, curl-command, extracted-results, etc.;
- Adds a host and matcher check when merging similar Nuclei findings to one defectdojo finding

**Test results**

Unit tests have been revised and extended.

**Documentation**

No need to make changes to the documentation.

**Checklist**

This checklist is for your information.

- [x] Make sure to rebase your PR against the very latest `dev`.
- [x] Features/Changes should be submitted against the `dev`.
- [x] Give a meaningful name to your PR, as it may end up being used in the release notes.
- [x] Your code is flake8 compliant.
- [x] Your code is python 3.11 compliant.
- [x] Add applicable tests to the unit tests.
- [x] Add the proper label to categorize your PR.

### CODE REVIEW

**Comment by Maffooch** [comment]
@kir-b this looks more like a rewrite than an improvement. Unit test scan files look pretty different as well. Did the format of the nuclei report change? If so, is it possible to maintain both versions?

**Comment by damiencarol** [comment]
@Maffooch to do what you're asking (supporting multiple version of the report) we need to not touch the old reports, exactly what this PR is doing :D 

@kir-b can you change this PR **to not touch old reports and add new ones**.

**Review by damiencarol — CHANGES_REQUESTED** [review]
This is a good PR that add the missing configuration part for de-duplication but you must not touch old reports.
Please create new unit tests and add new report instead of modifying old ones by hand.

**Comment by kir-b** [comment]
@Maffooch Yes, parser has changed a lot. Recent versions of Nuclei produce a slightly different reports than scan examples in unittest. But parser support both versions, I have tested. I also combined the results of all versions into one file (unittests/scans/nuclei/many_findings.json), but not everything that was before.
@damiencarol I can change this PR to not touch old reports and add new ones. But in this case, I still have to change the unit tests for the old reports. The reason is that I [added additional `matcher` and `endpoint.host` keys](https://github.com/DefectDojo/django-DefectDojo/pull/8057/files#diff-45bf85c6f622b3d6a0301a654876b334925d4a1e89be29443141ba1d0762e0b4R128) when merging multiple Nuclei finds into a single DefectDojo find. This is necessary because the current parser merges too many results into one finding and some of the information may be lost. This way I can revert old scans, add new ones, and rewrite old and new unit tests. Is it ok?

**Comment by hvechtomov** [comment]
I can confirm that the problem exists with new version of Nuclei #8060

**Comment by hvechtomov** [comment]
@damiencarol @Maffooch re-check please. This issue is quite critical, due to `nuclei` popularity. Thanks

**Comment by Maffooch** [comment]
@kir-b please make changes requested by @damiencarol

**Comment by kir-b** [comment]
@Maffooch @damiencarol Done. Please review.

File: dojo/tools/nuclei/parser.py [hunk 1]
```diff
@@ -22,63 +29,112 @@ def get_findings(self, filename, test):
...
         if len(data) == 0:
             return []
-        else:
-            dupes = {}
-            for item in data:
+
+        dupes = {}
+        for item in data:
+            logger.debug('Item %s.', str(item))
+            if item.get('templateID'):
```
Comment (Maffooch) [inline dojo/tools/nuclei/parser.py:36]:
```suggestion
            template_id = item.get('templateID', item.get('template-id', ''))
```

File: dojo/tools/nuclei/parser.py [hunk 2]
```diff
@@ -22,63 +29,112 @@ def get_findings(self, filename, test):
...
+            if severity not in Finding.SEVERITIES:
+                logger.debug('Unsupported severity value "%s", change to "%s"',
+                    severity, self.DEFAULT_SEVERITY)
+                severity = self.DEFAULT_SEVERITY
+            item_type = item.get('type')
+            if item_type is None:
+                item_type = ''
+            matched = item.get('matched')
+            if matched is None:
+                matched = item.get('matched-at')
```
Comment (Maffooch) [inline dojo/tools/nuclei/parser.py:54]:
```suggestion
            matched = item.get('matched', item.get('matched-at', ''))
```

File: dojo/tools/nuclei/parser.py [hunk 3]
```diff
@@ -22,63 +29,112 @@ def get_findings(self, filename, test):
...
+                    finding.cvssv3_score = classification['cvss-score']
+
+            if item.get('matcher-name'):
+                matcher = item.get('matcher-name')
+                finding.component_name = matcher
+            elif item.get('matcher_name'):
+                matcher = item.get('matcher_name')
+                finding.component_name = matcher
+            else:
+                matcher = ''
```
Comment (Maffooch) [inline dojo/tools/nuclei/parser.py:108]:
```suggestion
            matcher = item.get('matcher-name', item.get('matcher_name'))
            if matcher:
                finding.component_name = matcher
```

File: dojo/tools/nuclei/parser.py [hunk 4]
```diff
@@ -22,63 +29,112 @@ def get_findings(self, filename, test):
...
+            if info.get('remediation'):
+                finding.mitigation = info.get('remediation')
+
+            host = item.get('host')
+            if host is None:
+                host = ''
```
Comment (Maffooch) [inline dojo/tools/nuclei/parser.py:114]:
```suggestion
            host = item.get('host', '')
```

**Review by Maffooch — CHANGES_REQUESTED** [review]
Looking a lot better! Few nitpicky things

**Review by mtesauro — APPROVED** [review]
Approved

**Review by devGregA — APPROVED** [review]

**Review by cneill — APPROVED** [review]

**Comment by mtesauro** [comment]
@damiencarol Are you OK with this?

**Review by damiencarol — APPROVED** [review]

**Comment by damiencarol** [comment]
all good  @mtesauro  !

**Pull request merged**